Mathlib/Order/CompleteLatticeIntervals.lean

```diff
@@ -178,3 +179,49 @@ noncomputable instance ordConnectedSubsetConditionallyCompleteLinearOrder [Inhab
 #align ord_connected_subset_conditionally_complete_linear_order ordConnectedSubsetConditionallyCompleteLinearOrder
 
 end OrdConnected
+
+section Icc
+
+/-- Complete lattice structure on `Set.Icc` -/
+noncomputable def Set.Icc.completeLattice [ConditionallyCompleteLattice α]
+    {a b : α} (h : a ≤ b) : CompleteLattice (Set.Icc a b) :=
+{ Set.Icc.boundedOrder h with
+  sSup := fun S ↦ if hS : S = ∅ then ⟨a, le_rfl, h⟩ else ⟨sSup ((↑) '' S), by
+    rw [←Set.not_nonempty_iff_eq_empty, not_not] at hS
+    refine' ⟨_, csSup_le (hS.image (↑)) (fun _ ⟨c, _, hc⟩ ↦ hc ▸ c.2.2)⟩
+    obtain ⟨c, hc⟩ := hS
+    exact c.2.1.trans (le_csSup ⟨b, fun _ ⟨d, _, hd⟩ ↦ hd ▸ d.2.2⟩ ⟨c, hc, rfl⟩)⟩
+  le_sSup := by
+    intro S c hc
+    by_cases hS : S = ∅ <;> simp only [hS, dite_true, dite_false]
+    · simp [hS] at hc
+    · exact le_csSup ⟨b, fun _ ⟨d, _, hd⟩ ↦ hd ▸ d.2.2⟩ ⟨c, hc, rfl⟩
+  sSup_le := by
+    intro S c hc
+    by_cases hS : S = ∅ <;> simp only [hS, dite_true, dite_false]
+    · exact c.2.1
+    · exact csSup_le ((Set.nonempty_iff_ne_empty.mpr hS).image (↑))
+        (fun _ ⟨d, h, hd⟩ ↦ hd ▸ hc d h)
+  sInf := fun S ↦ if hS : S = ∅ then ⟨b, h, le_rfl⟩ else ⟨sInf ((↑) '' S), by
```

Comment:
Can you add a lemma that says on nonempty sets, `(sInf s : α) = sInf ((↑) '' S)`, and similarly for `iInf` and the `Sup`s?

Reply:
How should this be phrased? The issue that I'm having is that `sInf S` is not defined since the lattice structure on `Set.Icc a b` is a def rather than an instance.

Reply:
You can phrase it as:
```lean
lemma coe_sInf (h) :
    letI := thisInstanceDef h
    the_statement_now_that_the_instance_is_available :=by
  letI := thisInstanceDef h  -- we need this again in the proof
  sorry
```